src/main/java/org/folio/service/transactions/OpenTransactionsServiceImpl.java

```diff
@@ -0,0 +1,82 @@
+  public static final String MANUAL_BLOCKS = "manualblocks";
+  public static final String PROXIES = "proxies";
+  private CirculationStorageModuleClient circulationClient;
+  private FeesFinesModuleClient feesFinesClient;
+  private UserModuleClient userClient;
+
+  public OpenTransactionsServiceImpl(CirculationStorageModuleClient circulationClient, FeesFinesModuleClient feesFinesClient, UserModuleClient userClient) {
+    this.circulationClient = circulationClient;
+    this.feesFinesClient = feesFinesClient;
+    this.userClient = userClient;
+  }
+
+  @Override
+  public Future<OpenTransactions> getTransactionsOfUser(User user, OkapiConnectionParams connectionParams) {
+    String id = user.getId();
+    Future<Map.Entry<String, Integer>> loansFuture =
+      circulationClient.getOpenLoansCountByUserId(id, connectionParams)
+        .map(numberLoans -> Map.entry(LOANS, numberLoans));
+
+    Future<Map.Entry<String, Integer>> requestsFuture =
+      circulationClient.getOpenRequestsCountByUserId(id, connectionParams)
+        .map(numberReqs -> Map.entry(REQUESTS, numberReqs));
+
+    Future<Map.Entry<String, Integer>> accountsFuture =
+      feesFinesClient.getOpenAccountsCountByUserId(id, connectionParams)
+        .map(accounts -> Map.entry(ACCOUNTS, accounts));
+
+    Future<Map.Entry<String, Integer>> manualBlockFuture =
+      feesFinesClient.getManualBlocksCountByUserId(id, connectionParams)
+        .map(manualblocks -> Map.entry(MANUAL_BLOCKS, manualblocks));
+
+    Future<Map.Entry<String, Integer>> proxiesFuture =
+      userClient.getProxiesCountByUserId(id, connectionParams)
+        .map(p -> Map.entry(PROXIES, p));
+
+    Promise<OpenTransactions> result = Promise.promise();
+    CompositeFuture.all(loansFuture, requestsFuture, accountsFuture, manualBlockFuture, proxiesFuture)
+      .onSuccess(asyncResult -> {
+        CompositeFuture compFutResult = asyncResult.result();
+        Map<String, Integer> userTransactions = compFutResult.list().stream()
+          .map(o -> (Map.Entry<String, Integer>) o)
+          .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+        boolean hasOpenTransactions = compFutResult.list().stream()
+          .map(o -> (Map.Entry<String, Integer>) o)
+          .anyMatch(entry -> entry.getValue() > 0);
+        OpenTransactions openTransactions = new OpenTransactions()
+          .withLoans(userTransactions.get(LOANS))
+          .withRequests(userTransactions.get(REQUESTS))
+          .withBlocks(userTransactions.get(MANUAL_BLOCKS))
+          .withFeesFines(userTransactions.get(ACCOUNTS))
+          .withProxies(userTransactions.get(PROXIES))
+          .withHasOpenTransactions(hasOpenTransactions)
+          .withUserId(id);
+        if (user.getBarcode() != null) {
+          openTransactions.setUserBarcode(user.getBarcode());
+        }
+        result.complete(openTransactions);
+      })
+      .onFailure(result::fail);
+    return result.future();
+  }
```

Comment:
```
  @Override
  public Future<OpenTransactions> getTransactionsOfUser(User user, OkapiConnectionParams connectionParams) {
    String id = user.getId();
    final var openTransactions = new OpenTransactions().withUserId(id).withUserBarcode(user.getBarcode());

    Future<Integer> loansFuture =
      circulationClient.getOpenLoansCountByUserId(id, connectionParams)
        .onSuccess(numberLoans -> openTransactions.setLoans(numberLoans));

    Future<Integer> requestsFuture =
      circulationClient.getOpenRequestsCountByUserId(id, connectionParams)
        .onSuccess(numberReqs -> openTransactions.setRequests(numberReqs));

    Future<Integer> accountsFuture =
      feesFinesClient.getOpenAccountsCountByUserId(id, connectionParams)
        .onSuccess(accounts -> openTransactions.setFeesFines(accounts));

    Future<Integer> manualBlockFuture =
      feesFinesClient.getManualBlocksCountByUserId(id, connectionParams)
        .onSuccess(manualblocks -> openTransactions.setBlocks(manualblocks));

    Future<Integer> proxiesFuture =
      userClient.getProxiesCountByUserId(id, connectionParams)
        .onSuccess(p -> openTransactions.setProxies(p));

    return CompositeFuture.all(loansFuture, requestsFuture, accountsFuture, manualBlockFuture, proxiesFuture)
        .map(compositeFuture -> {
          boolean hasOpenTransactions = compositeFuture.result().list().stream()
              .map(o -> (Integer) o)
              .anyMatch(n -> n > 0);
          openTransactions.setHasOpenTransactions(hasOpenTransactions);
          return openTransactions;
        });
  }
```
Map.Entry stuff is not needed if each Future directly writes the result into openTransactions.
The Promise is not needed when directly returning CompositieFuture.all(...).map(...).
Calling setUserBarcode(null) or withUserBarcode(null) is fine and avoids the if clause.

Reply:
This is much cleaner. Thanks!